src/libappimage/utils/ElfFile.cpp

```diff
@@ -0,0 +1,151 @@
+// system
+extern "C" {
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+}
+
+// local
+#include "light_byteswap.h"
+#include "ElfFile.h"
+
+
+typedef Elf32_Nhdr Elf_Nhdr;
+
+
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+#define ELFDATANATIVE ELFDATA2LSB
+#elif __BYTE_ORDER == __BIG_ENDIAN
+#define ELFDATANATIVE ELFDATA2MSB
+#else
+#error "Unknown machine endian"
+#endif
+
+namespace appimage {
+    namespace utils {
+        ElfFile::ElfFile(const std::string& path) : path(path), fname(path.c_str()) {}
+
+        uint16_t ElfFile::file16_to_cpu(uint16_t val) {
+            if (ehdr.e_ident[EI_DATA] != ELFDATANATIVE)
+                val = bswap_16(val);
+            return val;
+        }
+
+        uint32_t ElfFile::file32_to_cpu(uint32_t val) {
+            if (ehdr.e_ident[EI_DATA] != ELFDATANATIVE)
+                val = bswap_32(val);
+            return val;
+        }
+
+        uint64_t ElfFile::file64_to_cpu(uint64_t val) {
+            if (ehdr.e_ident[EI_DATA] != ELFDATANATIVE)
+                val = bswap_64(val);
+            return val;
+        }
+
+        off_t ElfFile::read_elf32(FILE* fd) {
```

Comment:
The next two methods are redundant, they just differ a bit in types, called methods and maybe some numbers.